### PR TITLE
Update alert fix

### PR DESCRIPTION
**Pull request by riteshkarval**
- Update alert fix, update few conditions in multi condition alert. 
- Changing add_alert_condition function def, removed global prams to alert initialisation.

### CODE REVIEW

**Comment by riteshkarval** [comment]
Added delete alert condition support.

File: dkube/sdk/api.py [hunk 1]
```diff
@@ -2706,19 +2709,38 @@ def modelmonitor_update_alert(self, id, alert, alert_id):
 
         """
         current_alert = None
+        alert_class = None
+        alert_attr = lambda alert_class: "feature" if alert_class == "feature_drift" else "metric"
```
Comment (songole) [inline dkube/sdk/api.py]: There can be health alert as well?

Comment (riteshkarval) [reply dkube/sdk/api.py]: For health alerts also the key is used metric. Only feature drift alerts use feature as key.

File: dkube/sdk/api.py [hunk 2]
```diff
@@ -2706,19 +2709,38 @@ def modelmonitor_update_alert(self, id, alert, alert_id):
...
         if alert_id is None:
             raise ValueError("alert_id is recieved as None")
         existing_alerts = self.modelmonitor_get_alerts(id)
         for each_alert in existing_alerts:
             if each_alert["id"] == alert_id:
                 current_alert = each_alert
+                alert_class = each_alert["_class"]
```
Comment (songole) [inline dkube/sdk/api.py]: Is this possible?
I get the existing alert using some API. I update it myself and pass it as parameter for update.

Comment (riteshkarval) [reply dkube/sdk/api.py]: Yes, I tested it. It worked.

File: dkube/sdk/api.py [hunk 3]
```diff
@@ -2706,19 +2709,38 @@ def modelmonitor_update_alert(self, id, alert, alert_id):
...
         alert_dict = json.loads(alert.to_JSON())
-        for each_condition in alert_dict.get('conditions'):
-            if (not each_condition["threshold"] or (not each_condition["state"])):
-                alert_dict["conditions"] = current_alert["conditions"]
+        for each_condition in alert_dict['conditions']:
+            cond_existing = False
+            for each_current_condition in current_alert["conditions"]:
+                if each_current_condition[alert_key] == each_condition[alert_key]:
+                    cond_existing = True
+                    if len(each_condition) == 2: # delete condition will only have two items
```
Comment (songole) [inline dkube/sdk/api.py]: would this ever be successful? what does this expression return?

File: dkube/sdk/api.py [hunk 4]
```diff
@@ -2706,19 +2709,38 @@ def modelmonitor_update_alert(self, id, alert, alert_id):
...
+                alert_class = each_alert["_class"]
+        alert_key = alert_attr(alert_class)
         alert_dict = json.loads(alert.to_JSON())
-        for each_condition in alert_dict.get('conditions'):
-            if (not each_condition["threshold"] or (not each_condition["state"])):
-                alert_dict["conditions"] = current_alert["conditions"]
+        for each_condition in alert_dict['conditions']:
+            cond_existing = False
+            for each_current_condition in current_alert["conditions"]:
+                if each_current_condition[alert_key] == each_condition[alert_key]:
```
Comment (songole) [inline dkube/sdk/api.py]: alert_key & op would make it unique right?

File: dkube/sdk/api.py (hunk 2 above)
Comment (songole) [reply dkube/sdk/api.py]: looks like I don't understand this update API

File: dkube/sdk/api.py (hunk 4 above)
Comment (riteshkarval) [reply dkube/sdk/api.py]: Currently we can add multiple conditions having same alert_key and op. BE is accepting it. That’s why I kept only alert_key as key. And also user may update the op, in that case the comparison will fail if both are keys.

File: dkube/sdk/api.py (hunk 3 above)
Comment (riteshkarval) [reply dkube/sdk/api.py]: In delete the condition is having only two elements, whereas add condition have more than two elements.

File: dkube/sdk/api.py (hunk 4 above)
Comment (riteshkarval) [reply dkube/sdk/api.py]: Can we keep one alert condition per alert_key value for simplicity?

File: dkube/sdk/api.py [hunk 5]
```diff
@@ -2706,20 +2713,45 @@ def modelmonitor_update_alert(self, id, alert, alert_id):
...
+                    elif action == "delete":
+                        current_alert["conditions"].remove(each_current_condition)
+                    else:
+                        if each_condition["threshold"]:
+                            each_current_condition["threshold"] = each_condition["threshold"]
+                        if each_condition["state"]:
+                            each_current_condition["state"] = each_condition["state"]
+                        if each_condition["op"]:
+                            each_current_condition["op"] = each_condition["op"]
+            if (not cond_existing) and (action == "add"):
```
Comment (songole) [inline dkube/sdk/api.py]: What we discussed in the call doesn't work? They build DKubeModelmonitorAlert obj outside either using object methods or manually updating it. And, when they make update API, we just submit it as is. We can only choose to do validations. Is there a hole in this approach?

**Pull request merged**